central/processlisteningonport/datastore/datastore_impl.go

```diff
@@ -0,0 +1,503 @@
+		}
+
+		plopKey := getPlopKeyFromParts(val.GetProtocol(), val.GetPort(), indicatorID)
+
+		existingPLOP, prevExists := existingPLOPMap[plopKey]
+
+		if prevExists {
+			log.Debugf("Got existing PLOP to update timestamp: %+v", existingPLOP)
+
+			if existingPLOP.CloseTimestamp != nil &&
+				existingPLOP.CloseTimestamp != val.CloseTimestamp {
+
+				// An existing closed PLOP, update timestamp
+				existingPLOP.CloseTimestamp = val.CloseTimestamp
+				plopObjects = append(plopObjects, existingPLOP)
+			}
+
+			// Add nothing if the PLOP is active, i.e. CloseTimestamp == nil
+		}
+
+		if !prevExists {
+			if val.CloseTimestamp == nil {
+				// This events should always be closing by definition
+				log.Warnf("Found active PLOP completed in the batch %+v", val)
+			}
+
+			plopObjects = addNewPLOP(plopObjects, indicatorID, val.Process, val)
+		}
+	}
+
+	// Now save actual PLOP objects
+	err = ds.storage.UpsertMany(ctx, plopObjects)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ds *datastoreImpl) GetProcessListeningOnPort(
+	ctx context.Context,
+	deployment string,
+) (
+	processesListeningOnPorts []*storage.ProcessListeningOnPort, err error,
+) {
+	if ok, err := plopSAC.ReadAllowed(ctx); err != nil {
+		return nil, err
+	} else if !ok {
+		return nil, sac.ErrResourceAccessDenied
+	}
+
+	processesListeningOnPorts, err = ds.storage.GetProcessListeningOnPort(ctx, deployment)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return processesListeningOnPorts, nil
+}
+
+func (ds *datastoreImpl) WalkAll(ctx context.Context, fn WalkFn) error {
+	if ok, err := plopSAC.ReadAllowed(ctx); err != nil {
+		return err
+	} else if !ok {
+		return sac.ErrResourceAccessDenied
+	}
+
+	return ds.storage.Walk(ctx, fn)
+}
+
+func (ds *datastoreImpl) RemoveProcessListeningOnPort(ctx context.Context, ids []string) error {
+	if ok, err := plopSAC.WriteAllowed(ctx); err != nil {
+		return err
+	} else if !ok {
+		return sac.ErrResourceAccessDenied
+	}
+
+	return ds.removePLOP(ctx, ids)
+}
+
+func (ds *datastoreImpl) removePLOP(ctx context.Context, ids []string) error {
+
+	if len(ids) == 0 {
+		return nil
+	}
+
+	return ds.storage.DeleteMany(ctx, ids)
+}
+
+// fetchExistingPLOPs: Query already existing PLOP objects belonging to the
+// specified process indicators.
+//
+// XXX: This function queries all PLOP, no matter if they are matching port +
+// protocol we've got or not. This means potentially dangerous corner cases
+// when one process listens to a huge number of ports. To address it we could
+// introduce filtering by port and protocol to the query, and even without
+// extra indices PostgreSQL will be able to do it relatively efficiently using
+// bitmap scan.
+func (ds *datastoreImpl) fetchExistingPLOPs(
+	ctx context.Context,
+	indicatorIds []string,
+) (map[string]*storage.ProcessListeningOnPortStorage, error) {
+
+	var existingPLOPMap = map[string]*storage.ProcessListeningOnPortStorage{}
+
+	if len(indicatorIds) == 0 {
+		return nil, nil
```

Comment:
I'd rather return the empty map than nil here.

Reply:
Done